mutiny-core/src/indexed_db.rs

```diff
@@ -0,0 +1,248 @@
+            .await?;
+        Ok(mnemonic)
+    }
+
+    pub(crate) async fn get_mnemonic(&self) -> Result<Mnemonic, MutinyError> {
+        let tx = self
+            .indexed_db
+            .transaction(&[WALLET_OBJECT_STORE_NAME], TransactionMode::ReadOnly)?;
+        let store = tx.store(WALLET_OBJECT_STORE_NAME)?;
+
+        let key = JsValue::from(MNEMONIC_KEY);
+        let json = store.get(&key).await?;
+        let value: Option<String> = serde_wasm_bindgen::from_value(json)?;
+
+        let mnemonic = match value {
+            Some(mnemonic) => Mnemonic::from_str(&mnemonic)?,
+            None => return Err(MutinyError::InvalidMnemonic), // maybe need a better error
+        };
+
+        tx.done().await?;
+
+        Ok(mnemonic)
+    }
+
+    pub(crate) async fn has_mnemonic() -> Result<bool, MutinyError> {
+        let indexed_db = Self::build_indexed_db_database().await?;
+        let tx = indexed_db.transaction(&[WALLET_OBJECT_STORE_NAME], TransactionMode::ReadOnly)?;
+        let store = tx.store(WALLET_OBJECT_STORE_NAME)?;
+
+        let key = JsValue::from(MNEMONIC_KEY);
+        let json = store.get(&key).await?;
+        let value: Option<String> = serde_wasm_bindgen::from_value(json)?;
+
+        Ok(value.is_some())
+    }
+
+    async fn build_indexed_db_database() -> Result<Rexie, MutinyError> {
+        let rexie = Rexie::builder(WALLET_DATABASE_NAME)
+            .version(1)
+            .add_object_store(ObjectStore::new(WALLET_OBJECT_STORE_NAME))
+            .build()
+            .await?;
+
+        Ok(rexie)
+    }
+
+    async fn read_all(
+        indexed_db: &Rexie,
+        password: &Option<String>,
+    ) -> Result<HashMap<String, serde_json::Value>, MutinyError> {
+        let tx = indexed_db.transaction(&[WALLET_OBJECT_STORE_NAME], TransactionMode::ReadOnly)?;
+
+        let store = tx.store(WALLET_OBJECT_STORE_NAME)?;
+
+        let mut map = HashMap::new();
+        let all_json = store.get_all(None, None, None, None).await?;
+        let mut iter = all_json.into_iter();
+        while let Some((key, value)) = iter.next() {
+            let key = key
+                .as_string()
+                .ok_or(MutinyError::read_err(MutinyStorageError::Other(anyhow!(
+                    "key from indexedDB is not a string"
+                ))))?;
+            let json: serde_json::Value = match password {
+                Some(pw) if Self::needs_encryption(&key) => {
+                    let str: String = serde_wasm_bindgen::from_value(value)?;
+                    let ciphertext = decrypt(&str, pw);
+                    serde_json::from_str(&ciphertext)?
+                }
+                _ => serde_wasm_bindgen::from_value(value)?,
+            };
+            map.insert(key, json);
+        }
+
+        Ok(map)
+    }
+
+    fn needs_encryption(key: &str) -> bool {
+        // todo, what other keys need to be encrypted?
```

Comment:
probably just channel monitor? Might be worth asking the LDK team. I dont think channel mananger needs it so much.

But eventually probably everything if we're doing remote storage.

Reply:
asked

Reply:
Cool, either way this can be done in a follow up, good to leave that todo in there